Likelihood tempering for Stan PDF wrapper

simeoncarstens opened the pull request:
Adds support for likelihood tempering for the Stan PDF wrapper. This works by including additional "boolean" flags in the Stan model definition that allow switching on / off the likelihood / prior contributions. Credit for this idea goes to Stan developer @bob-carpenter (https://github.com/stan-dev/stan/issues/605#issuecomment-1348737675).

bob-carpenter commented:
You can also define tempered Stan models this way, but you either have to precompile a bunch of models, each with a copy of the data, or reload the model with new data dynamically.  Either way, the model can be defined this way.

```stan
functions {
  real log_likelihood(...) {
  }
  real log_prior(...) {
  ...
  }
}
data {
  real<lower=0, upper=1> t;
}
...
model {
  target += t * log_likelihood(...) + log_prior(...);  
}
```

Sorry we don't yet have a way to dynamically reset data or pin parameters to known values---either way would support tempering better.

simeoncarstens commented:
Thanks, that's good to know! What you propose seems quite a bit cleaner than what I implemented above. But I feel the `include_prior` / `exclude_likelihood` switches you suggested earlier and that I implemented allow to largely keep a standard Stan model definition and thus demands a bit less effort from the user when adapting the model to make it work with Chainsail's likelihood-tempering scheme.

Unless the answer to the following question is positive: in your code, in the `function`s, could one still somehow use the standard Stan `model` block syntax, such as `param ~ some_distribution(...)`? I only have a basic idea about writing Stan models.

As for supporting tempering in Stan directly, that's interesting. I'll follow that development! I don't know much about Sequential Monte Carlo, which you seem to be most interested in, but I'd be happy to chat about Parallel Tempering any time, if that's of interest to you and the other Stan developers.

bob-carpenter commented:
If you define a function `foo_lpdf(y, ...)`, then you can call `y ~ foo(...)`.  For example, 

```stan
functions {
  real foo_lpdf(real y, real theta) {
    return -0.5 * (y - theta)^2;
  }
}
parameters {
  real z;
}
model {
  z ~ foo(2.3);
}
```

So you could define a sampling distribution for the likelihood and a prior distribution for the prior and use the `lpdf` syntax.  Also, if the sampling distribution is discrete, then you need to use `lpmf`.
